docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,1119 @@
+
+- Including both generations (and the node ID) in object keys (see [object keys](#object-key-changes)) also provides some flexibility in
+  how HA should work in future: we could move to a model where failover can happen within one
+  attachment generation number (i.e. without control plane coordination) without changing the
+  storage format, because node IDs/generations would de-conflict writes from peers.
+- The node generation is also useful for managing remote objects which are not per-tenant,
+  such as the persistent per-node deletion queue which is proposed in this RFC.
+- node generation numbers enable building fencing into any network protocol (for example
+  communication with safekeepers) to refuse to communicate with stale/partitioned nodes.
+
+Note that the two generation numbers have a different behavior for stale generations:
+
+- A pageserver with a stale generation number should immediately terminate: it is never intended
+  to have two pageservers with the same node ID. This is not necessary for correctness, as the
+  node generation appears in object keys, but objects written in a stale node generation are never
+  read later, so waste storage.
+- An attachment with a stale generation number is permitted to continue operating (ingesting WAL
+  and serving reads), but not to do any deletions. This enables a multi-attached state for tenants,
+  facilitating live migration (this will be articulated in the next RFC that describes HA and
+  migrations). Operating in a stale attachment generation is safe, although for efficiency reasons
+  should not be done over long periods of time, as the data in the stale generation has to avoid
+  doing operations involving deletion, like remote compaction.
+
+### Object Key Changes
+
+#### Generation suffix
+
+All object keys (layer objects and index objects) will contain a numeric suffix that
+advances as the attachment generation and node generation advance.
+This suffix is the primary mechanism for protecting against split-brain situations, and
+enabling safe multi-attachment of tenants:
+
+- Two pageservers running with the same node ID (e.g. after a failure, where there is
+  some rogue pageserver still running) will not try to write to the same objects.
+- Multiple attachments of the same tenant will not try to write to the same objects, as
+  each attachment would have a distinct attachment generation.
+
+To avoid coupling our storage format tightly with the way we manage generation numbers,
+they will be packed into a single u64 that must obey just the following properties:
+
+- On a node restart, the suffix must increase
+- On a change to the attachment, the suffix must increase
+- Sorting the suffixes numerically must give the most logically recent data ([visibility](#visibility))
+
+```
+000000 0000 000000
+^^^^^^--------------------- 24 bit attachment generation
+
+       ^^^^---------------- 16 bit node ID
+
+            ^^^^^^--------- 24 bit node generation
+```
+
```

Comment:
Do we need the "Node ID" here at all? If I understand correctly, the concept of a node ID isn't really needed by this proposal.

Reply:
No, node ID went in here as an opportunistic thing in case we wanted to do some failover feature where the control plane wasn't in the loop, but we could always add it later if we needed to.

The suffix is now changed to be just the attachment generation.